File: frontend/src/api/monitoring.ts
/**
 * DB Monitoring API 封装
 *
 * 通过 Wails 生成的绑定调用后端监控方法
 */

// 直接调用 window.go.backend.App 上的方法
const go = () => (window as any)['go']['backend']['App']

// ─── TypeScript Interfaces ────────────────────────────────────────────────────

export interface ExecutionLogEntry {
  id: number
  timestamp: string
  connectionId: string
  dbType: 'mysql' | 'mongodb'
  database: string
  collection: string
  queryText: string
  executionTime: number
  rowsAffected: number
  success: boolean
}

export interface SlowQueryEntry {
  id: number
  timestamp: string
  connectionId: string
  dbType: 'mysql' | 'mongodb'
  database: string
  collection: string
  queryText: string
  durationMs: number
  rowsAffected: number
  errorMessage: string
}

export interface DataPoint {
  timestamp: string
  // common
  qps: number
  tps: number
  // MySQL
  threadsConnected: number
  threadsRunning: number
  innodbBufHitRate: number    // 0-100 %
  innodbRowLockWaits: number
  innodbBufPoolReads: number
  // MongoDB
  mongoConnections: number
  mongoPageFaults: number
  mongoMemResident: number    // MB
  mongoGlobalLock: number
}

export interface MonitoringSnapshot {
  profileId: string
  dbType: string
  dataPoints: DataPoint[]
}

// ─── API Functions ────────────────────────────────────────────────────────────

export function GetExecutionLog(profileID: string, dbType: string, limit: number): Promise<ExecutionLogEntry[]> {
  return go().GetExecutionLog(profileID, dbType, limit)
}

export function ClearExecutionLog(profileID: string, dbType: string): Promise<void> {
  return go().ClearExecutionLog(profileID, dbType)
}

export function GetMonitoringSnapshot(profileID: string, dbType: string): Promise<MonitoringSnapshot> {
  return go().GetMonitoringSnapshot(profileID, dbType)
}

export function GetSlowQueryLog(profileID: string, limit: number): Promise<SlowQueryEntry[]> {
  return go().GetSlowQueryLog(profileID, limit)
}

export function ClearSlowQueryLog(profileID: string): Promise<void> {
  return go().ClearSlowQueryLog(profileID)
}

export function SetSlowQueryThreshold(profileID: string, thresholdMs: number): Promise<void> {
  return go().SetSlowQueryThreshold(profileID, thresholdMs)
}

export function GetSlowQueryThreshold(profileID: string): Promise<number> {
  return go().GetSlowQueryThreshold(profileID)
}

File: backend/types/monitoring.go
package types

import "time"

// ExecutionLogEntry — unified view over query_history + mongo_query_history
type ExecutionLogEntry struct {
	ID            int       `json:"id"`
	Timestamp     time.Time `json:"timestamp"`
	ConnectionID  string    `json:"connectionId"`
	DBType        string    `json:"dbType"` // "mysql" | "mongodb"
	Database      string    `json:"database"`
	Collection    string    `json:"collection"`    // MongoDB only
	QueryText     string    `json:"queryText"`     // SQL or pipeline JSON
	ExecutionTime int64     `json:"executionTime"` // ms
	RowsAffected  int64     `json:"rowsAffected"`
	Success       bool      `json:"success"`
}

// SlowQueryEntry — row in slow_query_log
type SlowQueryEntry struct {
	ID           int       `json:"id"`
	Timestamp    time.Time `json:"timestamp"`
	ConnectionID string    `json:"connectionId"`
	DBType       string    `json:"dbType"`
	Database     string    `json:"database"`
	Collection   string    `json:"collection"`
	QueryText    string    `json:"queryText"`
	DurationMs   int64     `json:"durationMs"`
	RowsAffected int64     `json:"rowsAffected"`
	ErrorMessage string    `json:"errorMessage"`
}

// DataPoint is a single metrics sample (MySQL or MongoDB)
type DataPoint struct {
	Timestamp time.Time `json:"timestamp"`

	// Common
	QPS float64 `json:"qps"`
	TPS float64 `json:"tps"`

	// MySQL-specific
	ThreadsConnected   int64   `json:"threadsConnected"`   // Threads_connected
	ThreadsRunning     int64   `json:"threadsRunning"`     // Threads_running
	InnodbBufHitRate   float64 `json:"innodbBufHitRate"`   // buffer pool hit rate (0-100)
	InnodbRowLockWaits int64   `json:"innodbRowLockWaits"` // Innodb_row_lock_waits (cumulative)
	InnodbBufPoolReads int64   `json:"innodbBufPoolReads"` // Innodb_buffer_pool_reads (physical reads)

	// MongoDB-specific
	MongoConnections int64   `json:"mongoConnections"` // connections.current
	MongoPageFaults  int64   `json:"mongoPageFaults"`  // extra_info.page_faults
	MongoMemResident int64   `json:"mongoMemResident"` // mem.resident (MB)
	MongoGlobalLock  float64 `json:"mongoGlobalLock"`  // globalLock.currentQueue.total
}

// MonitoringSnapshot is the rolling window returned to the frontend
type MonitoringSnapshot struct {
	ProfileID  string      `json:"profileId"`
	DBType     string      `json:"dbType"` // "mysql" | "mongodb"
	DataPoints []DataPoint `json:"dataPoints"`
}
